### DISCUSSION TITLE
Allow superuser to change password for user programatically

### DISCUSSION THREAD

**Discussion by vickorian** [fastapi-users/fastapi-users#1392] [Q&A] [answered]
So there's a way to programmatically create a user as shown [here.](https://fastapi-users.github.io/fastapi-users/latest/cookbook/create-user-programmatically/) But, i haven't found a way to have a super user programmatically change the password of another user. I work with a group of people where having users reset their own passwords would open them up to phishing and spear phishing. Training them in proper opsec will be dramatically more difficult and less secure than just having a set of people capable of allowing them access to the servers. 

I know it's possible via post, but i haven't found a way to do it in the same flow as my current set up. Any help would be greatly appreciated.

**Comment by hgalytoby** [comment, marked as answer]
https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py#L665

```py
import contextlib

from app.db import get_async_session, get_user_db
from app.schemas import UserCreate
from app.users import get_user_manager
from fastapi_users.exceptions import UserNotExists

get_async_session_context = contextlib.asynccontextmanager(get_async_session)
get_user_db_context = contextlib.asynccontextmanager(get_user_db)
get_user_manager_context = contextlib.asynccontextmanager(get_user_manager)


async def update_password(email: str, password: str):
    try:
        async with get_async_session_context() as session:
            async with get_user_db_context(session) as user_db:
                async with get_user_manager_context(user_db) as user_manager:
                    user = await user_manager.get_by_email(email)
                    new_password = user_manager.password_helper.hash(password)
                    return await user_manager.user_db.update(user, {"hashed_password": new_password})
    except UserNotExists:
        print(f"User {email} not exists")
```

**Reply by vickorian** [reply]
This gets the job done with the modification that you should import "UserNotExists" from fastapi_users.exceptions and put that in place of "UserAlreadyExists"

**Reply by hgalytoby** [reply]
I copied the code for adding a new user from the document, but I forgot to modify the exception handling at the end!

**Reply by vickorian** [reply]
one last thing, is you don't need the new_password variable if you just pass the "password" variable as it will automatically create the "hashed_password" value in the dictionary in the user_manager._update method. But your modified version directly invoking user_db from the user manager allows you to directly interface with the DB.

```py
import contextlib

from app.db import get_async_session, get_user_db
from app.schemas import UserCreate
from app.users import get_user_manager
from fastapi_users.exceptions import UserNotExists

get_async_session_context = contextlib.asynccontextmanager(get_async_session)
get_user_db_context = contextlib.asynccontextmanager(get_user_db)
get_user_manager_context = contextlib.asynccontextmanager(get_user_manager)


async def update_password(email: str, password: str):
    try:
        async with get_async_session_context() as session:
            async with get_user_db_context(session) as user_db:
                async with get_user_manager_context(user_db) as user_manager:
                    user = await user_manager.get_by_email(email)
                    return await user_manager._update(user, {"password": password})
    except UserNotExists:
        print(f"User {email} not exists")
```

**Reply by hgalytoby** [reply]
You can choose the method that suits you best.
